ff/serving/provider/postgres.go

```diff
@@ -0,0 +1,501 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	db "github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"strings"
+	"time"
+)
+
+// postgresTableItem stores the value of a resource and its type.
+// Allows storage of any type and simpler table creation
+type postgresTableItem struct {
+	Value    interface{} `json:"value"`
+	ItemType string      `json:"type"`
+}
+
+type postgresOfflineStore struct {
+	conn             *pgxpool.Pool
+	ctx              context.Context
+	materializations map[MaterializationID]*memoryMaterialization //Are these used?
+	trainingSets     map[ResourceID]trainingRows
+	BaseProvider
+}
+
+type PostgresConfig struct {
+	Host     string
+	Port     string
+	Username string
+	Password string
+	Database string
+}
+
+func sanitize(ident string) string {
+	return db.Identifier{ident}.Sanitize()
+}
+
+func (pg *PostgresConfig) Deserialize(config SerializedConfig) error {
+	err := json.Unmarshal(config, pg)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (pg *PostgresConfig) Serialize() ([]byte, error) {
+	conf, err := json.Marshal(pg)
+	if err != nil {
+		return nil, err
+	}
+	return conf, nil
+}
+
+func postgresOfflineStoreFactory(config SerializedConfig) (Provider, error) {
+	pg := PostgresConfig{}
+	if err := pg.Deserialize(config); err != nil {
+		return nil, errors.New("invalid postgres config")
+	}
+
+	store, err := NewPostgresOfflineStore(pg)
+	if err != nil {
+		return nil, err
+	}
+	return store, nil
+}
+
+// NewPostgresOfflineStore creates a connection to a postgres database
+// and initializes a table to track currently active Resource tables.
+func NewPostgresOfflineStore(pg PostgresConfig) (*postgresOfflineStore, error) {
+	url := fmt.Sprintf("postgres://%s:%s@%s:%s/%s", pg.Username, pg.Password, pg.Host, pg.Port, pg.Database)
+	ctx := context.Background()
+	conn, err := pgxpool.Connect(ctx, url) //Change this
+	if err != nil {
+		return nil, err
+	}
+	return &postgresOfflineStore{
+		conn: conn,
+		ctx:  ctx,
+	}, nil
+}
+
+func (store *postgresOfflineStore) getResourceTableName(id ResourceID) string {
+	return fmt.Sprintf("resource_%s", id.Name)
```

Comment:
Doesn't distinguish tables by variant or id them by type

Reply:
Appended type and variant to table names